File: packages/yakr-core/src/yakr_core/session.py
from __future__ import annotations

import time
from dataclasses import dataclass

from cryptography.hazmat.primitives import serialization

from yakr_core.crypto import derive_mailbox_secret, hkdf_derive, xchacha_decrypt, xchacha_encrypt
from yakr_core.delivery_profile import DeliveryProfile
from yakr_core.ephemeral import DEFAULT_BLOB_TTL_MS, enforce_message_ttl
from yakr_core.errors import DecryptError, DuplicateSeqError, MessageExpiredError, RekeyRequiredError
from yakr_core.hybrid_pq import needs_pq_rekey
from yakr_core.identity import Contact, Identity
from yakr_core.mailbox import MailboxTag, MailboxTagDeriver
from yakr_core.message import InnerMessage, OuterBlob, message_id
from yakr_core.privacy import decode_padded_plaintext, pad_plaintext


@dataclass
class EncryptedMessage:
    outer_blob: OuterBlob
    inner_message: InnerMessage
    msg_id: str
    mailbox_tag: MailboxTag
    padding_bytes: int = 0


class Session:
    def __init__(self, identity: Identity, contact: Contact) -> None:
        self.identity = identity
        self.contact = contact
        if contact.ratchet is None:
            raise ValueError("contact missing double ratchet state; re-pair required")

    @property
    def send_direction(self) -> str:
        return _direction(self.identity.name, self.contact.name)

    @property
    def recv_direction(self) -> str:
        return _direction(self.contact.name, self.identity.name)

    def mailbox_deriver(self, *, outbound: bool) -> MailboxTagDeriver:
        direction = self.send_direction if outbound else self.recv_direction
        secret = derive_mailbox_secret(self.contact.master_secret, direction)
        epoch_secs = 3600
        if self.contact.delivery_profile is not None:
            epoch_secs = self.contact.delivery_profile.mailbox_epoch_secs
        return MailboxTagDeriver(secret, epoch_secs=epoch_secs)

    def _message_aad(self, inner: InnerMessage) -> bytes:
        return aad_for_message(
            valid_until=inner.valid_until,
            conversation_id=inner.conversation_id,
            seq=inner.seq,
        )

    def _encrypt_inner(self, inner: InnerMessage) -> tuple[bytes, int]:
        raw = inner.to_bytes()
        padded, padding_bytes = pad_plaintext(raw, self.contact.privacy_mode)
        ratchet_payload = self.contact.ratchet.encrypt(padded)
        return ratchet_payload, padding_bytes

    def _outer_blob(self, ciphertext: bytes, tag: MailboxTag) -> OuterBlob:
        return OuterBlob(
            version=1,
            mailbox_tag=tag.tag,
            expires_at=int(time.time() * 1000) + DEFAULT_BLOB_TTL_MS,
            ciphertext=ciphertext,
        )

    def encrypt_text(self, body: str) -> EncryptedMessage:
        self._require_fresh_session()
        seq = self.contact.next_send_seq
        inner = InnerMessage.text(
            conversation_id=self.contact.conversation_id,
            sender_device_id=self.identity.device_id,
            seq=seq,
            body=body,
        )
        ciphertext, padding_bytes = self._encrypt_inner(inner)
        tag = self.mailbox_deriver(outbound=True).derive(self.send_direction)
        outer = self._outer_blob(ciphertext, tag)
        self.contact.next_send_seq += 1
        return EncryptedMessage(
            outer_blob=outer,
            inner_message=inner,
            msg_id=message_id(outer.ciphertext),
            mailbox_tag=tag,
            padding_bytes=padding_bytes,
        )

    def encrypt_receipt(self, delivered_message_id: str) -> EncryptedMessage:
        self._require_fresh_session()
        seq = self.contact.next_send_seq
        inner = InnerMessage.receipt(
            conversation_id=self.contact.conversation_id,
            sender_device_id=self.identity.device_id,
            seq=seq,
            message_id=delivered_message_id,
        )
        ciphertext, padding_bytes = self._encrypt_inner(inner)
        tag = self.mailbox_deriver(outbound=True).derive(self.send_direction)
        outer = self._outer_blob(ciphertext, tag)
        self.contact.next_send_seq += 1
        return EncryptedMessage(
            outer_blob=outer,
            inner_message=inner,
            msg_id=message_id(outer.ciphertext),
            mailbox_tag=tag,
            padding_bytes=padding_bytes,
        )

    def encrypt_profile(self, profile: DeliveryProfile) -> EncryptedMessage:
        self._require_fresh_session()
        seq = self.contact.next_send_seq
        inner = InnerMessage.profile(
            conversation_id=self.contact.conversation_id,
            sender_device_id=self.identity.device_id,
            seq=seq,
            profile_b64=profile.to_b64(),
        )
        ciphertext, padding_bytes = self._encrypt_inner(inner)
        tag = self.mailbox_deriver(outbound=True).derive(self.send_direction)
        outer = self._outer_blob(ciphertext, tag)
        self.contact.next_send_seq += 1
        return EncryptedMessage(
            outer_blob=outer,
            inner_message=inner,
            msg_id=message_id(outer.ciphertext),
            mailbox_tag=tag,
            padding_bytes=padding_bytes,
        )

    def decrypt_outer(self, outer: OuterBlob) -> InnerMessage:
        mode = self.contact.privacy_mode
        try:
            padded = self.contact.ratchet.decrypt(outer.ciphertext)
        except ValueError as exc:
            if "already received" in str(exc):
                raise DuplicateSeqError("duplicate message") from exc
            raise DecryptError("unable to decrypt blob") from exc
        except Exception as exc:
            raise DecryptError("unable to decrypt blob") from exc
        try:
            plaintext = decode_padded_plaintext(padded, mode)
        except ValueError as exc:
            raise DecryptError("invalid padded plaintext") from exc
        inner = InnerMessage.from_bytes(plaintext)
        if inner.conversation_id != self.contact.conversation_id:
            raise DecryptError("conversation mismatch")
        if inner.seq <= self.contact.last_recv_seq:
            raise DuplicateSeqError(f"duplicate seq {inner.seq}")
        try:
            enforce_message_ttl(inner.valid_until)
        except MessageExpiredError:
            raise
        self.contact.last_recv_seq = inner.seq
        return inner

    def _require_fresh_session(self) -> None:
        if needs_pq_rekey(
            hybrid=self.contact.hybrid_pq,
            session_started_at_ms=self.contact.session_started_at,
            messages_sent=self.contact.next_send_seq,
        ):
            raise RekeyRequiredError("PQ session rekey required")


def _direction(sender: str, recipient: str) -> str:
    return f"{sender}->{recipient}"


def device_storage_key(identity: Identity) -> bytes:
    from yakr_core.crypto import hkdf_derive

    private_bytes = identity.signing_private.private_bytes(
        encoding=serialization.Encoding.Raw,
        format=serialization.PrivateFormat.Raw,
        encryption_algorithm=serialization.NoEncryption(),
    )
    return hkdf_derive(private_bytes, b"yakr/v1.0/local-message-store")


def wrap_local_ciphertext(identity: Identity, outer_ciphertext: bytes) -> bytes:
    from yakr_core.crypto import xchacha_encrypt

    return xchacha_encrypt(device_storage_key(identity), outer_ciphertext)


def unwrap_local_ciphertext(identity: Identity, payload: bytes) -> bytes:
    from yakr_core.crypto import xchacha_decrypt

    return xchacha_decrypt(device_storage_key(identity), payload)

File: rust/yakr-core/src/session.rs
use std::time::{SystemTime, UNIX_EPOCH};

use yakr_crypto::derive_mailbox_secret;

use crate::ephemeral::{message_valid_until, DEFAULT_BLOB_TTL_MS};
use crate::error::YakrError;
use crate::hybrid_pq::needs_pq_rekey;
use crate::identity::{Contact, Identity};
use crate::mailbox::{MailboxTag, MailboxTagDeriver};
use crate::message::{message_id, OuterBlob};
pub use yakr_crypto::{InnerMessage, MessageType};
use crate::privacy::{decode_padded_plaintext, pad_plaintext};
use crate::ratchet::RatchetState;

pub struct EncryptedMessage {
    pub outer_blob: OuterBlob,
    pub inner_message: InnerMessage,
    pub msg_id: String,
    pub mailbox_tag: MailboxTag,
    pub padding_bytes: usize,
}

pub struct Session {
    pub identity: Identity,
    pub contact: Contact,
}

impl Session {
    pub fn new(identity: Identity, mut contact: Contact) -> Result<Self, YakrError> {
        if contact.ratchet.is_none() {
            return Err(YakrError::Other(
                "contact missing double ratchet state; re-pair required".into(),
            ));
        }
        Ok(Self { identity, contact })
    }

    fn send_direction(&self) -> String {
        direction(&self.identity.name, &self.contact.name)
    }

    fn recv_direction(&self) -> String {
        direction(&self.contact.name, &self.identity.name)
    }

    fn mailbox_deriver(&self, outbound: bool) -> MailboxTagDeriver {
        let direction = if outbound {
            self.send_direction()
        } else {
            self.recv_direction()
        };
        let secret = derive_mailbox_secret(&self.contact.master_secret, &direction);
        MailboxTagDeriver::new(secret, self.contact.mailbox_epoch_secs)
    }

    fn require_fresh_session(&self) -> Result<(), YakrError> {
        if needs_pq_rekey(
            self.contact.hybrid_pq,
            self.contact.session_started_at,
            self.contact.next_send_seq,
        ) {
            return Err(YakrError::RekeyRequired);
        }
        Ok(())
    }

    fn encrypt_inner(&mut self, inner: &InnerMessage) -> Result<(Vec<u8>, usize), YakrError> {
        let raw = inner.to_bytes().map_err(|e| YakrError::Other(e.to_string()))?;
        let mode = self.contact.privacy_mode.clone();
        let (padded, padding_bytes) =
            pad_plaintext(&raw, &mode).map_err(YakrError::Other)?;
        let ratchet = self
            .contact
            .ratchet
            .as_mut()
            .ok_or_else(|| YakrError::Other("missing ratchet".into()))?;
        let ciphertext = ratchet
            .encrypt(&padded)
            .map_err(|e| YakrError::Other(e))?;
        Ok((ciphertext, padding_bytes))
    }

    fn outer_blob(&self, ciphertext: Vec<u8>, tag: MailboxTag) -> OuterBlob {
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap_or_default()
            .as_millis() as u64;
        OuterBlob {
            version: 1,
            mailbox_tag: tag.tag,
            expires_at: now + DEFAULT_BLOB_TTL_MS,
            ciphertext,
        }
    }

    pub fn encrypt_text(&mut self, body: &str) -> Result<EncryptedMessage, YakrError> {
        self.require_fresh_session()?;
        let seq = self.contact.next_send_seq;
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap_or_default()
            .as_millis() as u64;
        let inner = InnerMessage::text(
            self.contact.conversation_id.clone(),
            self.identity.device_id(),
            seq,
            body,
            now,
            message_valid_until(now),
        );
        let (ciphertext, padding_bytes) = self.encrypt_inner(&inner)?;
        let tag = self
            .mailbox_deriver(true)
            .derive(&self.send_direction(), None);
        let outer = self.outer_blob(ciphertext, tag.clone());
        self.contact.next_send_seq += 1;
        Ok(EncryptedMessage {
            msg_id: message_id(&outer.ciphertext),
            outer_blob: outer,
            inner_message: inner,
            mailbox_tag: tag,
            padding_bytes,
        })
    }

    pub fn decrypt_outer(&mut self, outer: &OuterBlob) -> Result<InnerMessage, YakrError> {
        let mode = self.contact.privacy_mode.clone();
        let ratchet_snapshot = self
            .contact
            .ratchet
            .as_ref()
            .map(|r| r.to_dict())
            .ok_or_else(|| YakrError::Other("missing ratchet".into()))?;
        let last_recv_before = self.contact.last_recv_seq;

        let padded = match self.contact.ratchet.as_mut().unwrap().decrypt(&outer.ciphertext) {
            Err(e) if e.contains("already received") => return Err(YakrError::DuplicateSeq),
            Err(e) => return Err(YakrError::Decrypt(e)),
            Ok(p) => p,
        };

        let plaintext = decode_padded_plaintext(&padded, &mode).map_err(|e| {
            self.restore_ratchet(&ratchet_snapshot);
            YakrError::Decrypt(e)
        })?;

        let inner = InnerMessage::from_bytes(&plaintext)
            .map_err(|e| {
                self.restore_ratchet(&ratchet_snapshot);
                YakrError::Decrypt(e.to_string())
            })?;

        if inner.conversation_id != self.contact.conversation_id {
            self.restore_ratchet(&ratchet_snapshot);
            return Err(YakrError::Decrypt("conversation mismatch".into()));
        }
        if inner.seq <= last_recv_before {
            self.restore_ratchet(&ratchet_snapshot);
            return Err(YakrError::DuplicateSeq);
        }
        if inner.seq != last_recv_before + 1 {
            self.restore_ratchet(&ratchet_snapshot);
            return Err(YakrError::DuplicateSeq);
        }
        if crate::ephemeral::enforce_message_ttl(inner.valid_until).is_err() {
            self.restore_ratchet(&ratchet_snapshot);
            return Err(YakrError::MessageExpired);
        }
        self.contact.last_recv_seq = inner.seq;
        Ok(inner)
    }

    fn restore_ratchet(&mut self, snapshot: &crate::ratchet::RatchetDict) {
        if let Ok(r) = RatchetState::from_dict(snapshot) {
            self.contact.ratchet = Some(r);
        }
    }

    pub fn into_contact(self) -> Contact {
        self.contact
    }
}

fn direction(sender: &str, recipient: &str) -> String {
    format!("{sender}->{recipient}")
}
